torch/distributed/_tensor/experimental/dmap.py

```diff
@@ -0,0 +1,115 @@
+from functools import wraps
+from typing import Callable, Optional, Sequence, Tuple, Union
+
+import torch
+from torch.distributed._tensor import DeviceMesh, DTensor, Placement
+
+from torch.utils._pytree import tree_flatten, tree_unflatten
+
+
+PlacementType = Sequence[Placement]
+
+InputLayouts = Union[PlacementType, Tuple]
+OutputLayouts = Union[PlacementType, Tuple[PlacementType, ...]]
+
+
+def _process_dtensor_inputs(flat_args, input_placements):
+    input_placements = (
+        [None] * len(flat_args) if input_placements is None else input_placements
+    )
+
+    flat_local_args = []
+    for arg, spec in zip(flat_args, input_placements):
+        if isinstance(arg, DTensor):
+            if spec is not None:
+                # TODO: see if we should just redistribute the DTensor here
+                assert (
+                    arg.placements == spec
+                ), "DTensor input placement does not match the input placement hints"
+            flat_local_args.append(arg.to_local())
+        elif isinstance(arg, torch.Tensor):
+            raise RuntimeError("dmap only support DTensor as input")
+        else:
+            assert (
+                spec is None
+            ), "Non-DTensor input should not have placement hints, please specify as None"
+            flat_local_args.append(arg)
+
+    return flat_local_args
+
+
+def dmap(
+    func: Callable,
+    mesh: DeviceMesh,
+    out_placements: OutputLayouts,
+    input_placements: Optional[InputLayouts] = None,
+):
+    """
+    dmap is distributed map, ``dmap(func)`` returns a function that maps the `func` over the
+    dtensor inputs to their local shards. This let user write functions in a local (per-device)
+    mode, instead of the DTensor "global" mode. i.e. we can write manual collectives
+    inside and mix the function call with DTensor operations.
+
+    Args:
+        func (Callable): A Python function that takes one or more arguments.
+            Must return one or more Tensors.
+        mesh (:class:`DeviceMesh`): DeviceMesh that encodes the mesh topology
```

Comment:
Can `mesh` be inferred from the DTensor inputs?

Reply:
good point yeah, I think we can remove this mesh and totally infer this from dtensor inputs